src/pkg/scan/dao/scan/vulnerability.go

```diff
@@ -0,0 +1,230 @@
+	GetForScanner(ctx context.Context, registrationUUID string) ([]*VulnerabilityRecord, error)
+	// DeleteForScanner deletes vulnerability records for a scanner
+	DeleteForScanner(ctx context.Context, registrationUUID string) (int64, error)
+	// DeleteForReport deletes vulnerability records for a report
+	DeleteForReport(ctx context.Context, reportUUID string) (int64, error)
+	// DeleteForDigests deletes vulnerability records for a provided list of digests
+	DeleteForDigests(ctx context.Context, digests ...string) (int64, error)
+	// GetRecordIdsForScanner gets record ids of vulnerability records for a scanner
+	GetRecordIdsForScanner(ctx context.Context, registrationUUID string) ([]int, error)
+}
+
+// NewVulnerabilityRecordDao returns a new dao to handle vulnerability data
+func NewVulnerabilityRecordDao() VulnerabilityRecordDao {
+	return &vulnerabilityRecordDao{}
+}
+
+type vulnerabilityRecordDao struct{}
+
+// Create creates new vulnerability record.
+func (v *vulnerabilityRecordDao) Create(ctx context.Context, vr *VulnerabilityRecord) (int64, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return 0, err
+	}
+	_, vrID, err := o.ReadOrCreate(vr, "CVEID", "RegistrationUUID", "Digest", "Package", "PackageVersion")
+
+	return vrID, err
+}
+
+// Delete  deletes a vulnerability record
+func (v *vulnerabilityRecordDao) Delete(ctx context.Context, vr *VulnerabilityRecord) error {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return err
+	}
+	_, err = o.Delete(vr, "CVEID", "RegistrationUUID")
+
+	return err
+}
+
+// List lists the vulnerability records with given query parameters.
+// Keywords in query here will be enforced with `exact` way.
+// If the registration ID (which = the scanner ID is not specified), the results
+// would contain duplicate records for a CVE depending upon the number of registered
+// scanners which individually store data about the CVE. In such cases, it is the
+// responsibility of the calling code to de-duplicate the CVE records or bucket them
+// per registered scanner
+func (v *vulnerabilityRecordDao) List(ctx context.Context, query *q.Query) ([]*VulnerabilityRecord, error) {
+	o, err := orm.FromContext(ctx)
+	if err != nil {
+		return nil, err
+	}
+	qt := o.QueryTable(new(VulnerabilityRecord))
+
+	if query != nil {
+		if len(query.Keywords) > 0 {
+			for k, v := range query.Keywords {
+				if vv, ok := v.([]interface{}); ok {
+					qt = qt.Filter(fmt.Sprintf("%s__in", k), vv...)
+					continue
+				}
+
+				qt = qt.Filter(k, v)
+			}
+		}
+
+		if query.PageNumber > 0 && query.PageSize > 0 {
+			qt = qt.Limit(query.PageSize, (query.PageNumber-1)*query.PageSize)
+		}
+	}
+
+	l := make([]*VulnerabilityRecord, 0)
+	_, err = qt.All(&l)
+
+	return l, err
+}
+
+// InsertForReport inserts a vulnerability record in the context of scan report
+func (v *vulnerabilityRecordDao) InsertForReport(ctx context.Context, reportUUID string, vr *VulnerabilityRecord) (int64, error) {
```

Comment:
Is there a way to demarcate RBDM transaction boundaries with this ORM? We may get inconsistent reports when some insert statements fail. I'm also wondering if the ORM support batch SQL statements?

Reply:
There is transaction support in Beego but I have not evaluated it and am not yet well versed with it. @heww  can help. Can we take this up in the next round? I know transaction is very critical, but given the number of tables we are updating the granularity or coarseness of the transaction is something we may have to think about

Reply:
There is transaction middleware for the HTTP handler (no effect for the legacy APIs), when the API response is not successful (status code not in [200, 400)), the transaction will rollback, so the DAO which gets the ormer from the context usually not need to deal transaction.

Reply:
@heww What about webhook handler where we call UpdateReport method that initiates SQL inserts into multiple tables. Is that context also transactional?

Reply:
In harbor 2.2, we refactor the scan/scan all to the task/execution model. The webhook (it's named checkin in jobservice, jobservice send a post request to the core) from the jobservice to core component also migrated to the new API handler. The `context.Context` parameter passed to the `UpdateReport` method contains an ormer which will execution SQLs in a transaction.

Reply:
Thank you @heww for clarifying that bit!